### PR TITLE
Prevent project publishing for new projects and accounts

### PR DESCRIPTION
**Pull request by bencodeorg**
This PR prevents publishing of new projects (created within past 30 minutes) and publishing by new accounts (account created within past 7 days). These change intend to limit the amount of low quality/effort content published to the public gallery (see spec for more details).

The changes here include:
- UI updates:
  - the **share dialog** you see when hitting the "Share" button in the header will disable the publish button and notify users if their account / project is too new to save
  - the **publish dialog** that appears just before you publish a project to tell why you couldn't save if you try. We pop up this dialog directly (ie, without first showing the share dialog) when publishing directly from your project list at /projects and at the end of hour of code tutorials when you hit the "Finish" button
- Add Rails route and model logic to manage who can publish projects
- Update project routing middleware to allow use of Rails logic mentioned above around project saving (new pattern!)

Note that there is an exception made for projects created as part of Hour of Code tutorials, as well as for student accounts that were added to a section within the past year.

**Share dialog for a new project created by a new student (shows individual messages as appropriate)**

image

**Publish dialog when publishing fails (via /projects -- shows error message as appropriate)**

image

## Links

- spec: [Limit who can publish to the public gallery](https://docs.google.com/document/d/12ibTD9FxkRZiWjo9_wwl8CyzImr-Dlzw270ef9SlpOU/edit#heading=h.dltzor8mqxmn)
- jira ticket: [SL-1028](https://codedotorg.atlassian.net/browse/SL-1028)

## Testing story

Added new tests to cover the new controller behavior and model logic. I manually tested the following scenarios (looking at both the share dialog and the publish dialog):

- (override) Hour of Code tutorial (hello world): with an old student account that's not in a section, can share a newly created project
- (override) student in section: can share newly created standalone project
- project too new: can't share a new standalone project with signed in, old student account that's not in a section
- user too new: can't share a 5 hour old standalone applab project with a signed in new account that's not in a section
- restricted share mode: can't share project in restricted share mode
- signed in/not: I couldn't find a scenario where you can publish as a signed out user

### CODE REVIEW

File: dashboard/legacy/middleware/helpers/projects.rb [hunk 1]
```diff
@@ -13,12 +13,23 @@ class NotFound < Sinatra::NotFound
...
   def initialize(storage_id)
     @storage_id = storage_id
 
     @table = Projects.table
   end
 
+  #### NOTE: This references the Rails model (Project, singular)
+  #### rather than this middleware class (Projects, plural)
+  #### such that we can make use of model associations managed by Rails.
+  def get_rails_project(project_id)
```
Comment (bencodeorg) [inline dashboard/legacy/middleware/helpers/projects.rb:28]: Especially curious for thoughts on this -- I think now that the `projects` table is in the dashboard database, we can start moving over project logic (and routing) into more standard Rails approaches. I'm hoping that this is a good starting point for that.

**Comment by bencodeorg** [comment]
Sorry, I need to fix up some UI tests first before review. Awaiting some clarification from product before I can finish these up. Will add another when I've done that and this is fully ready for review!

File: apps/src/code-studio/components/ShareAllowedDialog.jsx [hunk 2]
```diff
@@ -343,22 +385,25 @@ class ShareAllowedDialog extends React.Component {
...
-                      disabled={!hasThumbnail}
-                      className="no-mc"
-                    >
-                      <span>{i18n.publish()}</span>
-                    </Button>
-                  )}
-                  {publishAllowed && isPublished && (
+                  {showPublishInfo &&
+                    (isLoadingAccountAndProjectAge ? (
+                      <Spinner size="medium" style={{marginRight: 16}} />
```
Comment (molly-moen) [inline apps/src/code-studio/components/ShareAllowedDialog.jsx]: nit: if possible, can we move inline styles into the `styles` object?

File: apps/src/code-studio/components/ShareAllowedDialog.jsx [hunk 3]
```diff
@@ -438,6 +483,30 @@ class ShareAllowedDialog extends React.Component {
                     </span>
                   </div>
                 )}
+                {showPublishInfo &&
+                  !isLoadingAccountAndProjectAge &&
+                  !isAccountOldEnoughToPublish && (
+                    <div style={{clear: 'both', marginTop: 10}}>
```
Comment (molly-moen) [inline apps/src/code-studio/components/ShareAllowedDialog.jsx]: nit: this div and the one below are the same except the text, can we factor into a helper?

**Review by molly-moen — APPROVED** [review]
This looks good to me, I have a couple code clean up suggestions that could be done as follow-ups if we need this in asap

File: apps/src/code-studio/components/ShareAllowedDialog.jsx (hunk 3 above)
Comment (sanchitmalhotra126) [reply apps/src/code-studio/components/ShareAllowedDialog.jsx]: 👍 Pulled out these two as well as other similar warnings into a helper

**Pull request merged**